### PR TITLE
SA9004: time equality check

### PR DESCRIPTION
**Pull request by stuartcarnie**
Adds static check for reporting potentially problematic `time.Time` (in)equality comparisons

### CODE REVIEW

File: staticcheck/lint.go [hunk 1]
```diff
@@ -1187,6 +1188,29 @@ func (c *Checker) CheckDiffSizeComparison(j *lint.Job) {
...
+				if binop.Op != token.EQL && binop.Op != token.NEQ {
+					continue
+				}
+
+				if types.TypeString(binop.X.Type(), nil) != "time.Time" ||
+					types.TypeString(binop.Y.Type(), nil) != "time.Time" {
+					continue
+				}
+
+				j.Errorf(binop, "using == or != operators to compare time is problematic; prefer Time.Equal")
```
Comment (dominikh) [inline staticcheck/lint.go]: Reword the message to read `should use time.Time.Equal when comparing time.Time values`

File: cmd/staticcheck/docs/checks/SA9004 [hunk 2]
```diff
@@ -0,0 +1 @@
+Using == or != operators to compare time.Time is problematic; prefer Time.Equal
```
Comment (dominikh) [inline cmd/staticcheck/docs/checks/SA9004]: Wrap operators and identifiers in backticks

Comment (dominikh) [inline cmd/staticcheck/docs/checks/SA9004]: Say `should use time.Time.Equal instead` instead of `prefer Time.Equal`

Comment (dominikh) [inline cmd/staticcheck/docs/checks/SA9004]: Needs a paragraph explaining *why* time.Time.Equal should be used instead, but I'll write that myself.

File: staticcheck/testdata/CheckTimeEqualityComparison.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+)
+
+func fn1() {
+	var t1, t2 time.Time
+
+	_ = t1 == t2 // MATCH /using == or != operators to compare time is problematic/
+	_ = t1 != t2 // MATCH /using == or != operators to compare time is problematic/
+}
+
+func fn2() {
```
Comment (dominikh) [inline staticcheck/testdata/CheckTimeEqualityComparison.go]: Drop the tests involving pointers. People shouldn't be passing around pointers to times like this.

**Comment by shurcooL** [comment]
Is this check viable? Will it skip cases where `==` operator must be used, e.g., when `time.Time` is a map key?

See #47 for previous discussion of this check (before Go 1.9 changes were finalized).

**Comment by dominikh** [comment]
> Is this check viable?

No, it is not. Thanks for linking to prior discussion that I was utterly incapable of finding myself.

**Comment by stuartcarnie** [comment]
Good discussion – is this PR considered closed, @dominikh?

**Comment by dominikh** [comment]
I'm afraid so.

**Pull request closed without merging**